### PR TITLE
haskell: default name to "" when root isn't a path in developPackage

### PR DESCRIPTION
**Pull request by expipiplus1**
See https://github.com/NixOS/nixpkgs/issues/103062

I don't think the documentation on developPackage really needs updating, at the risk of making it too verbose.

###### Motivation for this change

This allows derivation outputs to be used as the `root` attribute in `developPackage`

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 1]
```diff
@@ -230,7 +230,7 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
     # build that package.
     developPackage =
       { root
-      , name ? builtins.baseNameOf root
+      , name ? if builtins.typeOf root == "path" then builtins.baseNameOf root else ""
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/make-package-set.nix:233]: When would you want this?  `root` gets passed to `callCabal2nix`, so it looks like it has to be a path?

What bug are you trying to fix here?

**Comment by expipiplus1** [comment]
This allows one to set root to store paths, for instance the output of
gitignoreSource.

File: pkgs/development/haskell-modules/make-package-set.nix (hunk 1 above)
Comment (expipiplus1) [reply pkgs/development/haskell-modules/make-package-set.nix:233]: Ah, email response doesn't work for review comments. My response is below

**Comment by expipiplus1** [comment]
Is there a nice way to get the name of a store path (i.e. everything after the hash), perhaps that would be better than `""`. Either way that string is just cosmetic.

**Comment by expipiplus1** [comment]
rebased

**Comment by expipiplus1** [comment]
Is it possible to merge or reject this please?

**Comment by expipiplus1** [comment]
rebased

**Comment by expipiplus1** [comment]
Maralorn, you are such a tease.

**Comment by maralorn** [comment]
I am sorry. I just a had blip where a browser plugin accidentally removed the haskell-updates branch. I haven‘t looked into the content of this PR but I will, if you are still on it.

**Comment by expipiplus1** [comment]
I've rebased, would be good to have this in or out one way or the other :)

**Comment by maralorn** [comment]
Yeah, so I have looked into this. And I am certainly not opposed. But I have to say I really like the idea of getting some kind of identifier from the store path. I will look into that.

**Comment by maralorn** [comment]
Hm, we can handcraft that, but that would be no tiny amount of code and there is no convenience function that I can find.

**Comment by sternenseemann** [comment]
The more cosmetic solution to this would be using something like this: <https://cs.tvl.fyi/depot/-/blob/nix/utils/default.nix#L20-59>. Depending on how relevant the `name` it may not be worth it to enter the territory of `builtins.unsafe*` in order to solve solve the problem.

**Comment by maralorn** [comment]
> Thanks!

Thanks, to you!

**Pull request merged**